platform/ios/src/MGLMapView.mm

```diff
@@ -815,204 +879,239 @@ - (UIViewController *)viewControllerForLayoutGuides
-    BOOL useLayoutGuides = viewController.view && viewController.automaticallyAdjustsScrollViewInsets;
-    UIView *containerView = useLayoutGuides ? viewController.view : self;
-    
-    // compass view
-    //
-    [containerView removeConstraints:self.compassViewConstraints];
-    [self.compassViewConstraints removeAllObjects];
-    
-    if (useLayoutGuides) {
-        [self.compassViewConstraints addObject:
-         [NSLayoutConstraint constraintWithItem:self.compassView
-                                      attribute:NSLayoutAttributeTop
-                                      relatedBy:NSLayoutRelationGreaterThanOrEqual
-                                         toItem:viewController.topLayoutGuide
-                                      attribute:NSLayoutAttributeBottom
-                                     multiplier:1.0
-                                       constant:8.0]];
-    }
-    [self.compassViewConstraints addObject:
-     [NSLayoutConstraint constraintWithItem:self.compassView
-                                  attribute:NSLayoutAttributeTop
-                                  relatedBy:NSLayoutRelationGreaterThanOrEqual
-                                     toItem:self
-                                  attribute:NSLayoutAttributeTop
-                                 multiplier:1.0
-                                   constant:8.0 + self.contentInset.top]];
-    
-    [self.compassViewConstraints addObject:
-     [NSLayoutConstraint constraintWithItem:self
-                                  attribute:NSLayoutAttributeTrailing
-                                  relatedBy:NSLayoutRelationEqual
-                                     toItem:self.compassView
-                                  attribute:NSLayoutAttributeTrailing
-                                 multiplier:1.0
-                                   constant:8.0 + self.contentInset.right]];
+- (void)installConstraintsWithView:(UIView *)view
+                      constraints:(NSMutableArray *)constraints
+                         position:(MGLOrnamentPosition)position
+                             size:(CGSize)size
+                           offset:(CGPoint)offset {
+    NSMutableArray *updatedConstraints = [NSMutableArray array];
     
-    [containerView addConstraints:self.compassViewConstraints];
-    
-    // scale bar view
-    //
-    [containerView removeConstraints:self.scaleBarConstraints];
-    [self.scaleBarConstraints removeAllObjects];
-    
-    if (useLayoutGuides) {
-        [self.scaleBarConstraints addObject:
-         [NSLayoutConstraint constraintWithItem:self.scaleBar
-                                      attribute:NSLayoutAttributeTop
-                                      relatedBy:NSLayoutRelationGreaterThanOrEqual
-                                         toItem:viewController.topLayoutGuide
-                                      attribute:NSLayoutAttributeBottom
+    if (@available(iOS 11.0, *)) {
```

Comment:
This check is necessary to support the safe area, but unnecessarily limits our use of layout anchors (which are available starting in iOS 9, the minimum version our SDK supports) — adopting something like @frederoni proposed [here](
https://github.com/mapbox/mapbox-gl-native/pull/12147/files#diff-392c354c88b0e3a5e82a6da45d2c8716) should allow us to massively simplify this section (and remove the `else` entirely).

Reply:
@friedbunny Not only layout anchors, but the else also considered `topLayoutGuide` and `bottomLayoutGuide` of view controllers.

Reply:
@lloydsheng what Jason is asking here is that instead of having this massive if/else  we took the approach mentioned [here](https://github.com/mapbox/mapbox-gl-native/pull/12147/files#diff-392c354c88b0e3a5e82a6da45d2c8716) instead; where you have an attribute that is going to check what's te valid anchor:

```
- (NSLayoutYAxisAnchor *)mgl_safeTopAnchor {
    if (@available(iOS 11.0, *)) {
        return self.safeAreaLayoutGuide.topAnchor;
    } else {
        return self.topAnchor;
    }
}
```
to something like:
e.g.
```
[updatedConstraints addObject:[view.topAnchor constraintEqualToAnchor:self.mgl_safeTopAnchor constant:offset.y]];
```

With this change we could simplify the code below. Also what he asks is that since the main SDK removed support for iOS 8 you should remove this [code](https://github.com/mapbox/mapbox-gl-native/blob/50672849ff9b2f6c09a5be9380c5cd47cca31182/platform/ios/src/MGLMapView.mm#L915-L1048), although it may be added to the iOS 8 branches that your team keeps around.

@friedbunny please correct me if I'm wrong.